src/eos-updater-poll-common.c

```diff
@@ -305,14 +301,45 @@ get_booted_refspec (OstreeDeployment  *booted_deployment,
   return TRUE;
 }
 
-static GFile *
-get_refspec_to_upgrade_on_file_from_deployment (OstreeSysroot    *sysroot,
-                                                OstreeDeployment *booted_deployment)
+static gboolean
+get_refspec_to_upgrade_on_from_deployment (OstreeSysroot     *sysroot,
+                                           OstreeDeployment  *booted_deployment,
+                                           gchar            **out_refspec_to_upgrade_from_deployment,
+                                           GError           **error)
 {
-  g_autoptr(GFile) root = g_file_new_for_path ("/");
-  g_autoptr(GFile) deployment_directory = ostree_sysroot_get_deployment_directory (sysroot,
-                                                                                   booted_deployment);
-  return g_file_get_child (deployment_directory, REFSPEC_TO_UPGRADE_ON_PATH);
+  const gchar *checksum = ostree_deployment_get_csum (booted_deployment);
+  g_autoptr(GVariant) commit = NULL;
+  g_autoptr(GVariant) metadata = NULL;
+  g_autoptr(GVariant) ref_for_deployment_variant = NULL;
+  OstreeRepo *repo = NULL;
+
+  g_return_val_if_fail (out_refspec_to_upgrade_from_deployment != NULL, FALSE);
+
+  if (!ostree_sysroot_get_repo (sysroot, &repo, NULL, error))
+   return FALSE;
+
+  if (!ostree_repo_load_variant (repo,
+                                 OSTREE_OBJECT_TYPE_COMMIT,
+                                 checksum,
+                                 &commit,
+                                 error))
+    return FALSE;
+
+  /* Look up the refspec to upgrade from here */
+  metadata = g_variant_get_child_value (commit, 0);
+  ref_for_deployment_variant = g_variant_lookup_value (metadata,
+                                                       "eos.upgrade-ref-for-deployment",
```

Comment:
What happens if `eos.upgrade-ref-for-deployment` and `endoflife-rebase` are both set to conflicting things? As I said on Slack, I think it would be better if this was a boolean ‘boot into this one before continuing’, used in conjunction with `endoflife-rebase` which provides the ref name to switch to.

Reply:
Ehhh. I'm not sure if we can meaningfully do that without having to do a bunch of gymnastics in stepping back through history in the pull to find the relevant commit. We control the ostree build process so is `eos.upgrade-ref-for-deployment` and `endoflife-rebase` are set to conflicting things on the *same* commit, then to be honest that's the ostree builder's problem.

Reply:
> if `eos.upgrade-ref-for-deployment` and `endoflife-rebase` are set to conflicting things on the same commit, then to be honest that's the ostree builder's problem.

It’s definitely always better to prevent such situations ever being possible through the design of data structures, rather than having to remember to prohibit them everywhere that data structure is populated (which might eventually be more than the OSTree builder).

That said, @ramcq made the good point on Slack that if the new ref name is stored in `ostree.endoflife-rebase`, old versions of eos-updater will ignore the bit that says ‘reboot into this before proceeding’ and will greedily update past it. There was a bug in eos-updater meaning it wouldn’t follow two `eos.endoflife-rebase` commits in a row, so we could potentially hack our way past this problem that way; but I’m losing the will to fight this battle.

If upstream OSTree does standardise a way of doing checkpoint commits in future, they’re likely to base it around `ostree.endoflife-rebase`. If so, we can always add a checkpoint using our mechanism, then switch to the OSTree mechanism. So there is a graceful transition away from this.

Let’s go with `eos.upgrade-ref-for-deployment`. (Though maybe call it something shorter like `eos.checkpoint-target`?)

Reply:
I'm happy to go with `eos.checkpoint-target`